### PR TITLE
feat: Deep connector Jira (Read)

### LINKED ISSUE #709: Add Jira deep connector
- GetPostAuthInfo
- ListObjectMetadata
- Read
- Write

### PR DESCRIPTION
**Pull request by Cobalt0s**
Closes #709 

# Description

Read operation uses search API endpoint to return Jira issues.
The data returned from `issue endpoint` is not returned untouched.

Fields property contains data of interest, the gist of it.
Id property is manually attached. This is the only relevant property outside of fields.

# Testing

Manual script to retrieve list of issues.
image

### CODE REVIEW

File: atlassian/test/cloud2.json
Comment (Cobalt0s) [inline atlassian/test/cloud2.json:3]: This is what token introspeciton returns.
There are options of what container to choose. I employed the strategy of selecting container by workspace name. 
Id means CloudID.

File: atlassian/modules.go
Comment (Cobalt0s) [inline atlassian/modules.go:12]: @RajatPawar do you think the having a module with variables is a good approach?

File: atlassian/parse.go
Comment (Cobalt0s) [inline atlassian/parse.go:11]: I am not sure about general solution for response parsing.
There should be a dependency between Atlassian module and response format.

For now we make only 1 read call for "issues".

File: atlassian/connector.go
Comment (Cobalt0s) [inline atlassian/connector.go:56]: Module is a template awaiting for cloud id variable.
Connector hold this state to construct URLs. See methods below in this file.

**Review by Cobalt0s — COMMENTED** [review]
@RajatPawar ignore test files they are autogenerated.
Current code allows reading Jira issues via Search api call. 
(need to address lots of TODOs)

File: atlassian/modules.go
Comment (RajatPawar) [reply atlassian/modules.go:12]: Can this be a part of the base URL? I'd like us to do modules a little more comprehensively ideally. In this particular case, if it becomes a part of the base URL, it just is a workspace var and can behave like docusign.

Comment (Cobalt0s) [reply atlassian/modules.go:12]: I am struggling with balancing 2 things. Jira APIs and whatever should come in like Confluence. I would honestly introduce modules only after working on the second deep connector.

I was thinking that BaseURLs are shared among 2 while each has its own structure one could be `ex/jira` the other something else. Checking [Confluence docs](https://developer.atlassian.com/cloud/confluence/rest/v2/intro/#auth) it seems that base URL differs entierly.

I could keep `rest/api/3` as module for now.

Comment (Cobalt0s) [reply atlassian/modules.go:12]: It is a bit confusing to have `WithModule(.)` if there is only 1 or raw aka empty. Honestly, all connectors fall under this category. Only some will have `{ empty, defaultVersion, option[1...many] }.`

**Comment by Cobalt0s** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/amp-labs/connectors/710?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#710** <a href="https://app.graphite.dev/github/pr/amp-labs/connectors/710?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#712** <a href="https://app.graphite.dev/github/pr/amp-labs/connectors/712?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 1 other dependent PR ([#713](https://github.com/amp-labs/connectors/pull/713) <a href="https://app.graphite.dev/github/pr/amp-labs/connectors/713?utm_source=stack-comment-icon" target="_blank">Graphite</a>)
* **#711** <a href="https://app.graphite.dev/github/pr/amp-labs/connectors/711?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @Cobalt0s and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: common/jsonquery/converters.go
Comment (Cobalt0s) [inline common/jsonquery/converters.go:72]: This is a small utility function. 
When getting `{baseURL}/oauth/token/accessible-resources` the response is array `[]` instead of `{}`.
Thats the primary reason for this method. Existing methods to get array expect the key, while in this case it is "root", "dot".

File: providers/utils.go
Comment (Cobalt0s) [inline providers/utils.go:160]: This method was not used. I repurposed it and it came in handy.
I verified that server repo doesn't use it either.

File: atlassian/connector.go
Comment (jlimatampersand) [inline atlassian/connector.go:69]: In comment, can we have link to jira api doc where it explains the endpoint?

File: atlassian/parse.go
Comment (jlimatampersand) [inline atlassian/parse.go:24]: thanks for example. This is super helpful!

File: atlassian/parse.go
Comment (jlimatampersand) [inline atlassian/parse.go:17]: What do you mean records cannot be returned as is? Why is that?
Can you add the link to the doc as well?

Comment (jlimatampersand) [reply atlassian/parse.go:17]: If that is what the API is returning, then we should return as is since that is what the users will be expecting. @laurenzlong Please correct me if I am wrong.

**Review by jlimatampersand — COMMENTED** [review]
Overall looks good, but not sure about changing object schema. I think we shouldn't unless there is some requirements that I am not aware of. Please let me know if there is any.

All looks good, but I will leave the object transformation part to @laurenzlong.

**Review by jlimatampersand — APPROVED** [review]
I just had chat about flattening the data with @laurenzlong This is good.

**Comment by jlimatampersand** [comment]
But please do provide links to docs in the comments before merging this. Thanks.

File: atlassian/parse.go
Comment (Cobalt0s) [reply atlassian/parse.go:17]: I remember some connector long ago had alike problem, hence I went with such approach.

File: atlassian/parse.go
Comment (Cobalt0s) [reply atlassian/parse.go:24]: :pray:

**Pull request merged**
